Worker lib

alexmeli100 opened the pull request.

alexmeli100 committed:
Updated design.md

alexmeli100 committed:
Implemented worker library

alexmeli100 committed:
added basic tests for worker lib

alexmeli100 committed:
Merge branch 'master' into worker-lib

# Conflicts:
#	api/grpc/service.proto
#	design.md

alexmeli100 committed:
Merge branch 'master' into worker-lib

# Conflicts:
#	api/grpc/service.proto

r0mant commented on supervisor/process.go:
```diff
@@ -0,0 +1,411 @@
+		return "finished"
+	case Fatal:
+		return "fatal"
+	default:
```
Feels like Interrupted is missing from here.

alexmeli100 replied:
Thanks. This has been fixed

r0mant commented on supervisor/process.go:
```diff
@@ -0,0 +1,411 @@
+// setState sets the state of a process
+func (p *Process) setState(state ProcessState) {
+	// cannot the set the state of a finished process
+	if p.state == Finished || p.state == state {
+		return
+	}
```
Shouldn't this be under lock as well?

alexmeli100 replied:
This has been fixed as well

r0mant commented on supervisor/process.go:
```diff
@@ -0,0 +1,411 @@
+	defer p.stateLock.Unlock()
+
+	// the next state can only be Starting if the state is Initial
+	if p.state == Initial {
+		p.state = Starting
```
This implicit logic may be unexpected for the caller, why not just set the state to whatever is passed in (and have the caller pass proper state)?

alexmeli100 replied:
Yeah I thought some basic validation could help the caller avoid mistakes. This logic has been removed

r0mant commented on supervisor/process.go:
```diff
@@ -0,0 +1,411 @@
+
+// IsInitialState is true if process state is Initial
+func (p *Process) IsInitialState() bool {
+	p.stateLock.Lock()
+	defer p.stateLock.Unlock()
```
Could switch to r/w mutex instead and use RLock for operations that don't change the state.

alexmeli100 replied:
This has been fixed. Thanks

r0mant commented on supervisor/process.go:
```diff
@@ -0,0 +1,411 @@
+	p.status.Error = err
+	p.status.ExitCode = exitCode
+	p.status.StopTime = now.UnixNano()
+	p.setState(Finished)
```
This will override Interrupted status if it was set above?

alexmeli100 replied:
Yes it overrides the Interrupted status

Joerger replied:
What is the point of the `Interrupted` state then? Like `Starting`, it in immediately overwritten and never visible outside this function due to the lock.

awly replied:
is the Interrupted status needed then?

r0mant commented on supervisor/process.go:
```diff
@@ -0,0 +1,411 @@
+	p.Lock()
+	defer p.Unlock()
+
+	return &p.status
```
Return a copy by value instead.

alexmeli100 replied:
This has been fixed. Thanks

r0mant commented on supervisor/supervisor.go:
```diff
@@ -0,0 +1,314 @@
+// GetOutput returns the line buffer for the job
+func (s *Supervisor) GetOutput(id string) ([]LogOutput, error) {
+	s.access.Lock()
+	s.access.Unlock()
```
defer?

alexmeli100 replied:
This been fixed. Don't know how I missed this

r0mant commented on supervisor/process.go:
```diff
@@ -0,0 +1,411 @@
+		lastCharOffset := firstChar + newLineOffset
+
+		// check for carriage return and strip it if present
+		if newLineOffset > 0 && b[lastCharOffset-1] == '\r' {
```
Why is this needed?

alexmeli100 replied:
This is to handle carriage return in newlines but it might not be needed anymore since I will just stream the bytes

r0mant commented on supervisor/process.go:
```diff
@@ -0,0 +1,411 @@
+		var line string
+
+		// preprend any lines if buffer is not empty
+		if s.lastChar > 0 {
+			line = string(s.buf[0:s.lastChar])
+			s.lastChar = 0
+		}
+
+		line += string(b[firstChar:lastCharOffset])
```
All this logic is pretty complicated and looks pretty fragile TBH. Would things simplify if you didn't try to split the output by lines and just stream everything byte by byte?

alexmeli100 replied:
Yeah you are definitely right that it would be simpler to just stream the bytes and let the client handle the output

alexmeli100 committed:
Fixed some typos in comments

r0mant commented on supervisor/process.go:
```diff
@@ -0,0 +1,411 @@
+
+		line += string(b[firstChar:lastCharOffset])
+
+		s.streamChan <- line
```
Since you're pushing to the channel on write, how will the clients connecting to the job receive prior logs if the job doesn't produce output anymore?

alexmeli100 replied:
The `Supervisor` holds a buffer to store prior logs for each process. The client can always call `GetOutput()` to get the logs at any time.

r0mant commented on supervisor/supervisor.go:
```diff
@@ -0,0 +1,314 @@
+	}
+
+	// Wait for process to exit completely
+	<-job.Done()
```
When Stop is called over API, this will hold the client on the wire which is probably not a great UX.

alexmeli100 replied:
Thanks. This has been fixed. The client will be responsible for calling `Status()` to get the final status.

alexmeli100 committed:
Fixed bugs based on feedback.
 Added TODO to fix output streaming. The new implementation will stream bytes instead of lines

alexmeli100 committed:
Updated basic tests for Supervisor

alexmeli100 committed:
Fixed mutex in IsInitialState() to RLock

Joerger commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	default:
+		return "unknown"
+	}
+}
```
IMO a custom string type makes more sense here, Go iota's are rarely the best way to define constants.
```suggestion
type ProcessState string

const (
	Initial ProcessState = "intial"
	Starting ProcessState = "starting"
	Running ProcessState = "running"
	Stopping ProcessState = "stopping"
	Finished ProcessState = "finished"
	Interrupted ProcessState = "interrupted"
	Fatal ProcessState = "failed"
)
```

Joerger commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	streamStdErr chan string
+	stdOut       *StreamOutput
+	stdErr       *StreamOutput
+	state        ProcessState
```
state is duplicated in status.

Joerger commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	}
+
+	p.statusChan = make(chan Status, 1)
+	p.setState(Starting)
```
What is the purpose of the starting state? The process will always be either `Fatal` or `Running` at the end of the function. Since it's protected by a lock, the `Starting` state is never visible.

awly replied:
note: this state will never be seen
when this function returns, the state will always be changed to either Fatal or Running

Joerger commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	p.cmd.Stdout = p.stdOut
+	p.cmd.Stderr = p.stdErr
+
+	p.cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
```
These can be set in the `NewProcess` constructor.

Joerger commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+
+	p.cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+	now := time.Now()
+	if err := p.cmd.Start(); err != nil {
+		// process failed to start.
+		// exit immediately
+		p.status.StartTime = now.UnixNano()
+		p.status.StopTime = time.Now().UnixNano()
+		p.status.Error = err
+		p.setState(Fatal)
+		p.statusChan <- p.status
+		close(p.done)
+
+		return p.statusChan
+	}
+
+	p.status.StartTime = now.UnixNano()
```
```suggestion
	p.status.StartTime := time.Now().UnixNano()
	if err := p.cmd.Start(); err != nil {
		// process failed to start.
		// exit immediately
		p.status.StopTime = time.Now().UnixNano()
		p.status.Error = err
		p.setState(Fatal)
		p.statusChan <- p.status
		close(p.done)

		return p.statusChan
	}

```

Joerger commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	defer p.Unlock()
+
+	if p.IsInitialState() || p.IsFinalState() {
+		return nil
```
This should probably return an error like "Failed to stop process, already in Initial/Final state"

Joerger commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	defer p.Unlock()
+
+	if p.IsInitialState() || p.IsFinalState() {
+		return nil
```
This should probably return an error like "Failed to signal process, already in Initial/Final state"

Joerger commented on supervisor/supervisor.go:
```diff
@@ -0,0 +1,311 @@
+	_, err := exec.LookPath(cmd)
+
+	if err != nil {
+		msg := fmt.Sprintf("cannot add job, the system cannot find the given executable \"%s\"", cmd)
```
```suggestion
		msg := fmt.Sprintf("cannot start job, the system cannot find the given executable \"%s\"", cmd)
```

Joerger commented on supervisor/supervisor.go:
```diff
@@ -0,0 +1,311 @@
+		}
+	}(id, proc)
+
+	fs := <-sc
```
Is this the main use of the status chan? Could you just receive from `p.Done()`, lock p.status, and then read the status?

awly commented on design.md:
```diff
@@ -96,11 +96,14 @@ All TLS 1.3 cipher suites are supported
 * TLS_AES_256_GCM_SHA384
 * TLS_CHACHA20_POLY1305_SHA256
 * TLS_AES_128_GCM_SHA256
+* TLS_AES_128_CCM_8_SHA256
+* TLS_AES_128_CCM_SHA256
```
`crypto/tls` in Go does not support these
```suggestion
```

awly commented on design.md:
```diff
@@ -96,11 +96,14 @@ All TLS 1.3 cipher suites are supported
 
 ### Authorization
-When starting a job, the server assigns the job to the client using the common name of it's cerficate. A client can stop, get the status and get the output of a job it owns. Attempting to perform any of these operations for a job not owned by the client results in an `NOT_FOUND` error.
+When starting a job, the server assigns the job to the client using the common name of it's cerficate. A client can stop, get the status and get the output of a job it owns. Attempting to perform any of these operations for a job not owned by the client results in an `UNAUTHOURIZED` error.
```
Looks like this was reverted from the original design doc
```suggestion
When starting a job, the server assigns the job to the client using the common name of it's cerficate. A client can stop, get the status and get the output of a job it owns. Attempting to perform any of these operations for a job not owned by the client results in an `NOT_FOUDN` error.
```

Joerger commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	return syscall.Kill(-p.cmd.Process.Pid, sig)
+}
+
+// StdOut returns the standard output streaming channel
+func (p *Process) StdOut() chan string {
+	return p.streamStdOut
+}
+
+// StdErr returns the standard error streaming channel
+func (p *Process) StdErr() chan string {
+	return p.streamStdErr
+}
+
+// Done returnes a channel that is closed when the process finishes running
+func (p *Process) Done() chan struct{} {
+	return p.done
+}
```
These channels should be returned as read only.

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+)
+
+const (
+	DefaultLineBufferSize = 16384
```
curious: how did you choose this buffer size?

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	Stopping
+	Finished
+	Interrupted
+	Fatal
```
this looks different from https://github.com/alexmeli100/teleport-assessment/blob/master/api/grpc/service.proto#L52-L56
should the possible states be the same?

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	status       Status
+	statusChan   chan Status
+	done         chan struct{}
+	stateLock    *sync.RWMutex
```
what's the difference between the top mutex and `stateLock`?
is it possible to only keep one mutex and avoid potential deadlocks?

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+
+// Process is a wrapper around exec.Cmd to run linux processes
+type Process struct {
+	*sync.Mutex
```
nit: don't embed mutexes.
this allows the external caller to mess with it and cause unexpected behavior

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+		PID:        0,
+		Error:      nil,
+		ExitCode:   -1,
+		StopByUser: false,
```
nit: you can skip the zero values
```suggestion
		Cmd:        name + " " + a,
		ExitCode:   -1,
```

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	defer p.stateLock.RUnlock()
+
+	return p.state == Finished || p.state == Interrupted
+}
```
how about a `GetState` helper instead?

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+
+// Start starts the process and returns a channel the caller can
+// use to retrieve the final status
+func (p *Process) Start() <-chan Status {
```
it looks like the return channel here is only used by the `Supervisor` to wait for all jobs to terminate.
you could simplify the implementation by having a `Wait` method on the `Process` instead

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	Error      error
+	ExitCode   int
+	StartTime  int64
+	StopTime   int64
```
can these fields be `time.Time`?

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	// send the final status and close done channel to signal to
+	// goroutines the process has finished running
+	defer func() {
+		p.statusChan <- p.GetStatus()
```
this goroutine might leak if `statusChan` is not consumed anywhere

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+			err = nil
+
+			if status, ok := exitErr.Sys().(syscall.WaitStatus); ok {
+				exitCode = status.ExitStatus()
```
```suggestion
			exitCode = exitErr.ExitCode()
			if status, ok := exitErr.Sys().(syscall.WaitStatus); ok {
```

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+
+	if p.IsInitialState() || p.IsFinalState() {
+		return nil
+	}
```
```suggestion
	if p.state != Running {
		return nil
	}
```

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	p.setState(Stopping)
+	p.status.StopByUser = byUser
+
+	return syscall.Kill(-p.cmd.Process.Pid, syscall.SIGTERM)
```
```suggestion
	return p.cmd.Process.Signal(syscall.SIGTERM)
```

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+		p.setState(Stopping)
+	}
+
+	return syscall.Kill(-p.cmd.Process.Pid, sig)
```
No need to guess the state based on signal, it's not possible to predict what most signals will do
```suggestion
	if !p.state == Running {
		return nil
	}

	return p.cmd.Process.Signal(sig)
```

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,408 @@
+	return s
+}
+
+// TODO: stream bytes instead of lines
```
let's address this TODO now, it will massively simplify the code

awly reviewed (commented):
Note: didn't get to `supervisor.go` or tests yet

alexmeli100 committed:
Further Updated worker library based on feedback. Still TODO output streaming modifications

alexmeli100 committed:
Update design.md

Co-authored-by: Andrew Lytvynov <andrew@awly.dev>

alexmeli100 committed:
StreamOutput now streams bytes instead of lines

alexmeli100 committed:
Updated Process Stop

Co-authored-by: Andrew Lytvynov <andrew@awly.dev>

awly commented on supervisor/process.go:
```diff
@@ -0,0 +1,231 @@
+func (s *StreamOutput) Write(buf []byte) (int, error) {
+	n := len(buf)
+
+	for _, b := range buf {
+		s.streamChan <- b
+	}
```
making these channels be `chan []byte` would be much more efficient than sending 1 byte at a time

alexmeli100 replied:
Thanks. This has been fixed

awly commented on supervisor/supervisor.go:
```diff
@@ -0,0 +1,311 @@
+		return nil, errors.New(msg)
+	}
+
+	_, err := exec.LookPath(cmd)
```
note: this would not be needed if `NewProcess` started the command right away (without a separate `Start` method) and returned the error from `p.cmd.Start` instead of using an internal `error` field

alexmeli100 replied:
Thanks. This has been fixed. The Process now starts right away in `NewProcess`

alexmeli100 committed:
Streaming module now sends a byte slice instead of individual bytes

alexmeli100 committed:
Updated process to start command when creating it

alexmeli100 committed:
fixed output streaming implementation
